consensus/state.go

```diff
@@ -366,6 +367,25 @@ func (cs *State) OnStart() error {
 		return err
 	}
 
+	// Double Signing Risk Reduction
```

Comment:
Is it actually sufficient to only do this in `OnStart()`? If a node has e.g. fast sync disabled, or for whatever other reason isn't caught up with the chain, then a different validator can submit votes for heights that this node doesn't know about yet. You should be able to reproduce this simply by starting a new node with a duplicate key and fast sync disabled.

I believe we will have to run this check again before casting our first vote, after consensus has caught up with the chain head, or something similar.

Reply:
I agree, that would be a more secure way.
What do you think about adding check logic again in [`gossipDataRoutine`](https://github.com/tendermint/tendermint/blob/master/consensus/reactor.go#L540)(height and round matched) for that? and I could be set the `DoubleSignCheckHeight` as a flag to zero to make it work only once here

Reply:
I'm afraid I'm not very familiar with the intricacies of the consensus module, but I'd say the check should run right before we try to cast our first vote, probably in `State` somewhere, not `Reactor`.

Reply:
Hi. Erik. This is Hyung from B-Harvest. We are quite confident that the proposed additional check will satisfy raised concern about covering different route of risk.

Because this work has been staled,  could you elaborate more detail on the codebase about your concerns? It will be helpful to understand your opinion.

We will wait several days before we move on with our proposed approach.

Thank you for your suggestion!

Reply:
Yeah, it won't work without fast-sync / state-sync. We can create a separate issue for improving it. I doubt there's a clear point in code where "right before we try to cast our first vote" happens, so it might be difficult to pin point.

Reply:
sure, I'll create a separate issue for improving it after merged this fast-sync version